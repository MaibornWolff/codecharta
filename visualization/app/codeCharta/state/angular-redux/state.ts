import { Injectable } from "@angular/core"

import { Store as PlainStore } from "../store/store"

@Injectable({
	providedIn: "root"
})
export class State {
	getValue() {
		return PlainStore.store.getState()
	}
}
