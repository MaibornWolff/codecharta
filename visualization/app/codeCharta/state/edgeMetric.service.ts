import { MetricData, FileState, BlacklistItem, Edge, BlacklistType, CodeMapNode, EdgeMetricCount } from "../codeCharta.model"
import _ from "lodash"
import { FileStateServiceSubscriber, FileStateService } from "./fileState.service"
import { IRootScopeService } from "angular"
import { FileStateHelper } from "../util/fileStateHelper"
import { CodeMapHelper } from "../util/codeMapHelper"
import { BlacklistSubscriber } from "./settingsService/settings.service.events"
import { SettingsService } from "./settingsService/settings.service"
import { HierarchyNode } from "d3"

export interface EdgeMetricServiceSubscriber {
	onEdgeMetricDataUpdated(metricData: MetricData[])
}

export class EdgeMetricService implements FileStateServiceSubscriber, BlacklistSubscriber {
	private static EDGE_METRIC_DATA_UPDATED_EVENT = "edge-metric-data-updated"

	private edgeMetricData: MetricData[] = []
	private nodeEdgeMetricsMap: Map<string, Map<string, EdgeMetricCount>>

	constructor(private $rootScope: IRootScopeService, private fileStateService: FileStateService) {
		FileStateService.subscribe(this.$rootScope, this)
		SettingsService.subscribeToBlacklist(this.$rootScope, this)
	}

	public onBlacklistChanged(blacklist: BlacklistItem[]) {
		const fileStates: FileState[] = this.fileStateService.getFileStates()
		this.updateEdgeMetrics(fileStates, blacklist)
	}

	public onFileSelectionStatesChanged(fileStates: FileState[]) {
		this.updateEdgeMetrics(fileStates, [])
	}

	public onImportedFilesChanged(fileState: FileState[]) {}

	private updateEdgeMetrics(fileStates: FileState[], blacklist: BlacklistItem[]) {
		this.edgeMetricData = this.calculateMetrics(FileStateHelper.getVisibleFileStates(fileStates), blacklist)
		this.addNoneMetric()
		this.sortNodeEdgeMetricsMap()
		this.notifyEdgeMetricDataUpdated()
	}

	public getMetricNames(): string[] {
		if (!this.edgeMetricData) {
			return []
		}

		return this.edgeMetricData.map(x => x.name)
	}

	public getAmountOfAffectedBuildings(metricName: string): number {
		if (!this.nodeEdgeMetricsMap.has(metricName)) {
			return 0
		}
		return this.nodeEdgeMetricsMap.get(metricName).size
	}

	public getNodesWithHighestValue(metricName: string, numberOfNodes: number): string[] {
		if (!this.nodeEdgeMetricsMap.has(metricName)) {
			return []
		}

		let highestEdgeCountBuildings: string[] = []
		const edgeMetricMapKeyIterator = this.nodeEdgeMetricsMap.get(metricName).keys()
		for (let i = 0; i < numberOfNodes; i++) {
			highestEdgeCountBuildings.push(edgeMetricMapKeyIterator.next().value)
		}

		return highestEdgeCountBuildings
	}

	public getMetricValuesForNode(node: HierarchyNode<CodeMapNode>): Map<string, EdgeMetricCount> {
		const metricNames = this.getMetricNames().filter(it => !!this.nodeEdgeMetricsMap.get(it))
		let nodeEdgeMetrics = new Map()

		metricNames.forEach(metric => {
			nodeEdgeMetrics.set(metric, this.nodeEdgeMetricsMap.get(metric).get(node.data.path))
		})
		return nodeEdgeMetrics
	}

	public getMetricData(): MetricData[] {
		return this.edgeMetricData
	}

	private calculateMetrics(visibleFileStates: FileState[], blacklist: BlacklistItem[]): MetricData[] {
		if (visibleFileStates.length <= 0) {
			return []
		} else {
			const hashMap = this.calculateEdgeMetricData(visibleFileStates, blacklist)
			return this.getMetricDataFromMap(hashMap)
		}
	}

	private calculateEdgeMetricData(fileStates: FileState[], blacklist: BlacklistItem[]): Map<string, Map<string, EdgeMetricCount>> {
		this.nodeEdgeMetricsMap = new Map()
		fileStates.forEach(fileState => {
			fileState.file.settings.fileSettings.edges.forEach(edge => {
				if (this.bothNodesAssociatedAreVisible(edge, blacklist, fileStates)) {
					this.addEdgeToCalculationMap(edge)
				}
			})
		})
		return this.nodeEdgeMetricsMap
	}

	private bothNodesAssociatedAreVisible(edge: Edge, blacklist: BlacklistItem[], fileStates: FileState[]): boolean {
		const fromNode = this.getMapNodeFromPath(edge.fromNodeName, fileStates)
		const toNode = this.getMapNodeFromPath(edge.toNodeName, fileStates)
		return fromNode && toNode && this.isNotBlacklisted(fromNode, blacklist) && this.isNotBlacklisted(toNode, blacklist)
	}

	private getMapNodeFromPath(path: string, fileStates: FileState[]): CodeMapNode {
		let mapNode = undefined
		fileStates.forEach(fileState => {
			const nodeFound = CodeMapHelper.getCodeMapNodeFromPath(path, "File", fileState.file.map)
			if (nodeFound) {
				mapNode = nodeFound
			}
		})
		return mapNode
	}

	private isNotBlacklisted(node: CodeMapNode, blacklist: BlacklistItem[]): boolean {
		return !CodeMapHelper.isBlacklisted(node, blacklist, BlacklistType.exclude)
	}

	private addEdgeToCalculationMap(edge: Edge) {
		for (let edgeMetric of Object.keys(edge.attributes)) {
			const edgeMetricEntry = this.getEntryForMetric(edgeMetric)
			this.addEdgeToNodes(edgeMetricEntry, edge.fromNodeName, edge.toNodeName)
		}
	}

	private getEntryForMetric(edgeMetricName: string): Map<string, EdgeMetricCount> {
		if (!this.nodeEdgeMetricsMap.has(edgeMetricName)) {
			this.nodeEdgeMetricsMap.set(edgeMetricName, new Map())
		}
		return this.nodeEdgeMetricsMap.get(edgeMetricName)
	}

	private addEdgeToNodes(edgeMetricEntry: Map<string, EdgeMetricCount>, fromNode: string, toNode: string) {
		this.createEntryIfNecessary(edgeMetricEntry, fromNode)
		this.createEntryIfNecessary(edgeMetricEntry, toNode)
		edgeMetricEntry.get(fromNode).outgoing += 1
		edgeMetricEntry.get(toNode).incoming += 1
	}

	private createEntryIfNecessary(edgeMetricEntry: Map<string, EdgeMetricCount>, nodeName: string) {
		if (!edgeMetricEntry.has(nodeName)) {
			edgeMetricEntry.set(nodeName, { incoming: 0, outgoing: 0 })
		}
	}

	private getMetricDataFromMap(hashMap: Map<string, Map<string, EdgeMetricCount>>): MetricData[] {
		let metricData: MetricData[] = []

		hashMap.forEach((occurences: any, edgeMetric: any) => {
			let maximumMetricValue = 0
			occurences.forEach((value: EdgeMetricCount, _) => {
				const combinedValue = value.incoming + value.outgoing
				if (combinedValue > maximumMetricValue) {
					maximumMetricValue = combinedValue
				}
			})
			metricData.push({ name: edgeMetric, maxValue: maximumMetricValue, availableInVisibleMaps: true })
		})

		return metricData
	}

	private sortNodeEdgeMetricsMap() {
		let sortedEdgeMetricMap = new Map()
		this.nodeEdgeMetricsMap.forEach((value, key) => {
			const sortedMapForMetric = new Map(
				[...value.entries()].sort((a, b) => b[1].incoming + b[1].outgoing - (a[1].incoming + a[1].outgoing))
			)
			sortedEdgeMetricMap.set(key, sortedMapForMetric)
		})
		this.nodeEdgeMetricsMap = sortedEdgeMetricMap
	}

	private addNoneMetric() {
		this.edgeMetricData.push({ name: "None", maxValue: 0, availableInVisibleMaps: false })
	}

	private notifyEdgeMetricDataUpdated() {
		this.$rootScope.$broadcast(EdgeMetricService.EDGE_METRIC_DATA_UPDATED_EVENT, this.edgeMetricData)
	}

	public static subscribe($rootScope: IRootScopeService, subscriber: EdgeMetricServiceSubscriber) {
		$rootScope.$on(EdgeMetricService.EDGE_METRIC_DATA_UPDATED_EVENT, (event, data) => {
			subscriber.onEdgeMetricDataUpdated(data)
		})
	}
}
