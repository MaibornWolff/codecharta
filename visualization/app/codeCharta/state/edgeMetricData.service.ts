import { MetricData, BlacklistItem, Edge, BlacklistType, CodeMapNode, EdgeMetricCount, AttributeTypeValue } from "../codeCharta.model"
import { IRootScopeService } from "angular"
import { CodeMapHelper } from "../util/codeMapHelper"
import { HierarchyNode } from "d3"
import { BlacklistService, BlacklistSubscriber } from "./store/fileSettings/blacklist/blacklist.service"
import { FilesService, FilesSelectionSubscriber } from "./store/files/files.service"
import { Files } from "../model/files"
import { StoreService } from "./store.service"

export interface EdgeMetricDataServiceSubscriber {
	onEdgeMetricDataUpdated(metricData: MetricData[])
}

export class EdgeMetricDataService implements FilesSelectionSubscriber, BlacklistSubscriber {
	private static EDGE_METRIC_DATA_UPDATED_EVENT = "edge-metric-data-updated"

	private edgeMetricData: MetricData[] = []
	private nodeEdgeMetricsMap: Map<string, Map<string, EdgeMetricCount>>

	constructor(private $rootScope: IRootScopeService, private storeService: StoreService) {
		FilesService.subscribe(this.$rootScope, this)
		BlacklistService.subscribe(this.$rootScope, this)
	}

	public onBlacklistChanged(blacklist: BlacklistItem[]) {
		this.updateEdgeMetrics()
	}

	public onFilesSelectionChanged(files: Files) {
		this.updateEdgeMetrics()
	}

	private updateEdgeMetrics() {
		this.edgeMetricData = this.calculateMetrics()
		this.addNoneMetric()
		this.sortNodeEdgeMetricsMap()
		this.notifyEdgeMetricDataUpdated()
	}

	public getMetricNames(): string[] {
		if (!this.edgeMetricData) {
			return []
		}

		return this.edgeMetricData.map(x => x.name)
	}

	public getAmountOfAffectedBuildings(metricName: string): number {
		if (!this.nodeEdgeMetricsMap || !this.nodeEdgeMetricsMap.has(metricName)) {
			return 0
		}
		return this.nodeEdgeMetricsMap.get(metricName).size
	}

	public getNodesWithHighestValue(metricName: string, numberOfNodes: number): string[] {
		if (!this.nodeEdgeMetricsMap || !this.nodeEdgeMetricsMap.has(metricName)) {
			return []
		}

		let highestEdgeCountBuildings: string[] = []
		const edgeMetricMapKeyIterator = this.nodeEdgeMetricsMap.get(metricName).keys()
		for (let i = 0; i < numberOfNodes; i++) {
			highestEdgeCountBuildings.push(edgeMetricMapKeyIterator.next().value)
		}

		return highestEdgeCountBuildings
	}

	public getMetricValuesForNode(node: HierarchyNode<CodeMapNode>): Map<string, EdgeMetricCount> {
		const metricNames = this.getMetricNames().filter(it => !!this.nodeEdgeMetricsMap.get(it))
		let nodeEdgeMetrics = new Map()

		metricNames.forEach(metric => {
			nodeEdgeMetrics.set(metric, this.nodeEdgeMetricsMap.get(metric).get(node.data.path))
		})
		return nodeEdgeMetrics
	}

	public getMetricData(): MetricData[] {
		return this.edgeMetricData
	}

	public getAttributeTypeByMetric(metricName: string): AttributeTypeValue {
		return this.storeService.getState().fileSettings.attributeTypes.edges[metricName]
	}

	private calculateMetrics(): MetricData[] {
		if (!this.storeService.getState().files.fileStatesAvailable()) {
			return []
		} else {
			const hashMap = this.calculateEdgeMetricData()
			return this.getMetricDataFromMap(hashMap)
		}
	}

	private calculateEdgeMetricData(): Map<string, Map<string, EdgeMetricCount>> {
		this.nodeEdgeMetricsMap = new Map()
		const pathsPerFileState = this.storeService
			.getState()
			.files.getVisibleFileStates()
			.map(fileState => CodeMapHelper.getAllPaths(fileState.file.map))
		const allFilePaths: string[] = [].concat(...pathsPerFileState)
		this.storeService
			.getState()
			.files.getVisibleFileStates()
			.forEach(fileState => {
				fileState.file.settings.fileSettings.edges.forEach(edge => {
					if (this.bothNodesAssociatedAreVisible(edge, allFilePaths)) {
						this.addEdgeToCalculationMap(edge)
					}
				})
			})
		return this.nodeEdgeMetricsMap
	}

	private bothNodesAssociatedAreVisible(edge: Edge, filePaths: string[]): boolean {
		const fromPath = filePaths.find(x => x === edge.fromNodeName)
		const toPath = filePaths.find(x => x === edge.toNodeName)
		return fromPath && toPath && this.isNotBlacklisted(fromPath) && this.isNotBlacklisted(toPath)
	}

	private isNotBlacklisted(path: string): boolean {
		const blacklist = this.storeService.getState().fileSettings.blacklist
		return !CodeMapHelper.isPathBlacklisted(path, blacklist, BlacklistType.exclude)
	}

	private addEdgeToCalculationMap(edge: Edge) {
		for (let edgeMetric of Object.keys(edge.attributes)) {
			const edgeMetricEntry = this.getEntryForMetric(edgeMetric)
			this.addEdgeToNodes(edgeMetricEntry, edge.fromNodeName, edge.toNodeName)
		}
	}

	private getEntryForMetric(edgeMetricName: string): Map<string, EdgeMetricCount> {
		if (!this.nodeEdgeMetricsMap.has(edgeMetricName)) {
			this.nodeEdgeMetricsMap.set(edgeMetricName, new Map())
		}
		return this.nodeEdgeMetricsMap.get(edgeMetricName)
	}

	private addEdgeToNodes(edgeMetricEntry: Map<string, EdgeMetricCount>, fromNode: string, toNode: string) {
		this.createEntryIfNecessary(edgeMetricEntry, fromNode)
		this.createEntryIfNecessary(edgeMetricEntry, toNode)
		edgeMetricEntry.get(fromNode).outgoing += 1
		edgeMetricEntry.get(toNode).incoming += 1
	}

	private createEntryIfNecessary(edgeMetricEntry: Map<string, EdgeMetricCount>, nodeName: string) {
		if (!edgeMetricEntry.has(nodeName)) {
			edgeMetricEntry.set(nodeName, { incoming: 0, outgoing: 0 })
		}
	}

	private getMetricDataFromMap(hashMap: Map<string, Map<string, EdgeMetricCount>>): MetricData[] {
		let metricData: MetricData[] = []

		hashMap.forEach((occurences: any, edgeMetric: any) => {
			let maximumMetricValue = 0
			occurences.forEach((value: EdgeMetricCount, _) => {
				const combinedValue = value.incoming + value.outgoing
				if (combinedValue > maximumMetricValue) {
					maximumMetricValue = combinedValue
				}
			})
			metricData.push({ name: edgeMetric, maxValue: maximumMetricValue })
		})

		return metricData
	}

	private sortNodeEdgeMetricsMap() {
		let sortedEdgeMetricMap = new Map()
		if (this.nodeEdgeMetricsMap) {
			this.nodeEdgeMetricsMap.forEach((value, key) => {
				const sortedMapForMetric = new Map(
					[...value.entries()].sort((a, b) => b[1].incoming + b[1].outgoing - (a[1].incoming + a[1].outgoing))
				)
				sortedEdgeMetricMap.set(key, sortedMapForMetric)
			})
		}
		this.nodeEdgeMetricsMap = sortedEdgeMetricMap
	}

	private addNoneMetric() {
		this.edgeMetricData.push({ name: "None", maxValue: 0 })
	}

	private notifyEdgeMetricDataUpdated() {
		this.$rootScope.$broadcast(EdgeMetricDataService.EDGE_METRIC_DATA_UPDATED_EVENT, this.edgeMetricData)
	}

	public static subscribe($rootScope: IRootScopeService, subscriber: EdgeMetricDataServiceSubscriber) {
		$rootScope.$on(EdgeMetricDataService.EDGE_METRIC_DATA_UPDATED_EVENT, (event, data) => {
			subscriber.onEdgeMetricDataUpdated(data)
		})
	}
}
