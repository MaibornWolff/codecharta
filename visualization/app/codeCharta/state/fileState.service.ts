import { CCFile, FileSelectionState, FileState } from "../codeCharta.model"
import { IRootScopeService } from "angular"
import { LoadingGifService } from "../ui/loadingGif/loadingGif.service"

export interface FileStateServiceSubscriber {
	onFileSelectionStatesChanged(fileStates: FileState[])

	onImportedFilesChanged(fileStates: FileState[])
}

export class FileStateService {
	private static FILE_STATE_CHANGED_EVENT = "file-selection-states-changed"
	private static IMPORTED_FILES_CHANGED_EVENT = "imported-files-changed"

	private fileStates: Array<FileState> = []

	/* @ngInject */
	constructor(private $rootScope: IRootScopeService, private loadingGifService: LoadingGifService) {}

	public resetMaps() {
		this.fileStates = []
	}

	public addFile(file: CCFile) {
		this.fileStates.push({ file: file, selectedAs: FileSelectionState.None })
		this.notifyFileImport()
	}

	public getCCFiles(): CCFile[] {
		return this.fileStates.map(x => x.file)
	}

	public getFileStates(): FileState[] {
		return this.fileStates
	}

	public setSingle(file: CCFile) {
		this.resetSelectionStates()
		const matchedFile = this.fileStates.find(x => x.file === file)
		if (matchedFile) {
			matchedFile.selectedAs = FileSelectionState.Single
		}
		this.notifySelectionChange()
	}

	public setDelta(reference: CCFile, comparison: CCFile) {
		this.resetSelectionStates()
		const matchedReferenceFile = this.fileStates.find(x => x.file === reference)
		const matchedComparisonFile = this.fileStates.find(x => x.file === comparison)

		if (matchedReferenceFile) {
			matchedReferenceFile.selectedAs = FileSelectionState.Reference
		}

		if (matchedComparisonFile) {
			matchedComparisonFile.selectedAs = FileSelectionState.Comparison
		}
		this.notifySelectionChange()
	}

	public setMultiple(multipleFiles: CCFile[]) {
		this.resetSelectionStates()
		this.fileStates.filter(x => multipleFiles.indexOf(x.file) !== -1).forEach(x => (x.selectedAs = FileSelectionState.Partial))
		this.notifySelectionChange()
	}

	private resetSelectionStates() {
		this.fileStates.forEach(file => (file.selectedAs = FileSelectionState.None))
	}

	private notifySelectionChange() {
		this.loadingGifService.updateLoadingMapFlag(true)
		this.$rootScope.$broadcast(FileStateService.FILE_STATE_CHANGED_EVENT, this.fileStates)
	}

	private notifyFileImport() {
		this.$rootScope.$broadcast(FileStateService.IMPORTED_FILES_CHANGED_EVENT, this.fileStates)
	}

	public static subscribe($rootScope: IRootScopeService, subscriber: FileStateServiceSubscriber) {
		$rootScope.$on(FileStateService.FILE_STATE_CHANGED_EVENT, (event, data) => {
			subscriber.onFileSelectionStatesChanged(data)
		})
		$rootScope.$on(FileStateService.IMPORTED_FILES_CHANGED_EVENT, (event, data) => {
			subscriber.onImportedFilesChanged(data)
		})
	}
}
