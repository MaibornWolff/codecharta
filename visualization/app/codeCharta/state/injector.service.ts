// Plop: Append service import here
import { FilesService } from "./store/files/files.service"
import { SearchPanelModeService } from "./store/appSettings/searchPanelMode/searchPanelMode.service"
import { IsLoadingFileService } from "./store/appSettings/isLoadingFile/isLoadingFile.service"
import { IsLoadingMapService } from "./store/appSettings/isLoadingMap/isLoadingMap.service"
import { MapSizeService } from "./store/treeMap/mapSize/mapSize.service"
import { MapColorsService } from "./store/appSettings/mapColors/mapColors.service"
import { ResetCameraIfNewFileIsLoadedService } from "./store/appSettings/resetCameraIfNewFileIsLoaded/resetCameraIfNewFileIsLoaded.service"
import { ShowOnlyBuildingsWithEdgesService } from "./store/appSettings/showOnlyBuildingsWithEdges/showOnlyBuildingsWithEdges.service"
import { WhiteColorBuildingsService } from "./store/appSettings/whiteColorBuildings/whiteColorBuildings.service"
import { IsWhiteBackgroundService } from "./store/appSettings/isWhiteBackground/isWhiteBackground.service"
import { DynamicMarginService } from "./store/appSettings/dynamicMargin/dynamicMargin.service"
import { InvertHeightService } from "./store/appSettings/invertHeight/invertHeight.service"
import { InvertDeltaColorsService } from "./store/appSettings/invertDeltaColors/invertDeltaColors.service"
import { InvertColorRangeService } from "./store/appSettings/invertColorRange/invertColorRange.service"
import { HideFlatBuildingsService } from "./store/appSettings/hideFlatBuildings/hideFlatBuildings.service"
import { CameraService } from "./store/appSettings/camera/camera.service"
import { ScalingService } from "./store/appSettings/scaling/scaling.service"
import { EdgeHeightService } from "./store/appSettings/edgeHeight/edgeHeight.service"
import { AmountOfEdgePreviewsService } from "./store/appSettings/amountOfEdgePreviews/amountOfEdgePreviews.service"
import { AmountOfTopLabelsService } from "./store/appSettings/amountOfTopLabels/amountOfTopLabels.service"
import { MarkedPackagesService } from "./store/fileSettings/markedPackages/markedPackages.service"
import { EdgesService } from "./store/fileSettings/edges/edges.service"
import { AttributeTypesService } from "./store/fileSettings/attributeTypes/attributeTypes.service"
import { EdgeMetricService } from "./store/dynamicSettings/edgeMetric/edgeMetric.service"
import { ColorRangeService } from "./store/dynamicSettings/colorRange/colorRange.service"
import { MarginService } from "./store/dynamicSettings/margin/margin.service"
import { SearchPatternService } from "./store/dynamicSettings/searchPattern/searchPattern.service"
import { SearchedNodePathsService } from "./store/dynamicSettings/searchedNodePaths/searchedNodePaths.service"
import { FocusedNodePathService } from "./store/dynamicSettings/focusedNodePath/focusedNodePath.service"
import { HeightMetricService } from "./store/dynamicSettings/heightMetric/heightMetric.service"
import { DistributionMetricService } from "./store/dynamicSettings/distributionMetric/distributionMetric.service"
import { ColorMetricService } from "./store/dynamicSettings/colorMetric/colorMetric.service"
import { AreaMetricService } from "./store/dynamicSettings/areaMetric/areaMetric.service"
import { BlacklistService } from "./store/fileSettings/blacklist/blacklist.service"
import { NodeSearchService } from "./nodeSearch.service"
import { IsPresentationModeService } from "./store/appSettings/isPresentationMode/isPresentationMode.service"

export class InjectorService {
	/* @ngInject */
	constructor(
		// tslint:disable:no-unused-variable
		// We have to inject the services somewhere
		// Plop: Append service injection here
		private searchPanelModeService: SearchPanelModeService,
		private isLoadingFileService: IsLoadingFileService,
		private isLoadingMapService: IsLoadingMapService,
		private filesService: FilesService,
		private mapSizeService: MapSizeService,
		private mapColorsService: MapColorsService,
		private resetCameraIfNewFileIsLoadedService: ResetCameraIfNewFileIsLoadedService,
		private showOnlyBuildingsWithEdgesService: ShowOnlyBuildingsWithEdgesService,
		private whiteColorBuildingsService: WhiteColorBuildingsService,
		private isWhiteBackgroundService: IsWhiteBackgroundService,
		private dynamicMarginService: DynamicMarginService,
		private invertHeightService: InvertHeightService,
		private invertDeltaColorsService: InvertDeltaColorsService,
		private invertColorRangeService: InvertColorRangeService,
		private hideFlatBuildingsService: HideFlatBuildingsService,
		private cameraService: CameraService,
		private scalingService: ScalingService,
		private edgeHeightService: EdgeHeightService,
		private amountOfEdgePreviewsService: AmountOfEdgePreviewsService,
		private amountOfTopLabelsService: AmountOfTopLabelsService,
		private markedPackagesService: MarkedPackagesService,
		private edgesService: EdgesService,
		private attributeTypesService: AttributeTypesService,
		private edgeMetricService: EdgeMetricService,
		private colorRangeService: ColorRangeService,
		private marginService: MarginService,
		private searchPatternService: SearchPatternService,
		private searchedNodePathsService: SearchedNodePathsService,
		private focusedNodePathService: FocusedNodePathService,
		private heightMetricService: HeightMetricService,
		private distributionMetricService: DistributionMetricService,
		private colorMetricService: ColorMetricService,
		private areaMetricService: AreaMetricService,
		private blacklistService: BlacklistService,
		private nodeSearchService: NodeSearchService,
		private isPresentationModeService: IsPresentationModeService
	) {}
}
