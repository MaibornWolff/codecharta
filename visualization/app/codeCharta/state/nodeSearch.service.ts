import { CodeMapNode } from "../codeCharta.model"
import { IRootScopeService } from "angular"
import { CodeMapHelper } from "../util/codeMapHelper"
import { CodeMapPreRenderService } from "../ui/codeMap/codeMap.preRender.service"
import { StoreService } from "./store.service"
import { setSearchedNodePaths } from "./store/dynamicSettings/searchedNodePaths/searchedNodePaths.actions"
import { SearchPatternService, SearchPatternSubscriber } from "./store/dynamicSettings/searchPattern/searchPattern.service"

export interface NodeSearchSubscriber {
	onNodeSearchComplete(searchedNodes: CodeMapNode[])
}

export class NodeSearchService implements SearchPatternSubscriber {
	private static NODE_SEARCH_COMPLETE_EVENT = "node-search-complete"

	private searchedNodes: CodeMapNode[] = []

	/* @ngInject */
	constructor(
		private $rootScope: IRootScopeService,
		private storeService: StoreService,
		private codeMapPreRenderService: CodeMapPreRenderService
	) {
		SearchPatternService.subscribe(this.$rootScope, this)
	}

	onSearchPatternChanged(searchPattern: string) {
		this.searchedNodes = CodeMapHelper.getNodesByGitignorePath(this.codeMapPreRenderService.getRenderMap(), searchPattern)
		this.notifyNodeSearchComplete()
		this.applySettingsSearchedNodePaths()
	}

	private applySettingsSearchedNodePaths() {
		const newSearchedNodePaths = this.searchedNodes.map(x => x.path)
		this.storeService.dispatch(setSearchedNodePaths(new Set(newSearchedNodePaths)))
	}

	private notifyNodeSearchComplete() {
		this.$rootScope.$broadcast(NodeSearchService.NODE_SEARCH_COMPLETE_EVENT, this.searchedNodes)
	}

	static subscribe($rootScope: IRootScopeService, subscriber: NodeSearchSubscriber) {
		$rootScope.$on(NodeSearchService.NODE_SEARCH_COMPLETE_EVENT, (_event_, data) => {
			subscriber.onNodeSearchComplete(data)
		})
	}
}
