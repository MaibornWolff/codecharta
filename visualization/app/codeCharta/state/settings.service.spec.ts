import "./state.module"
import {SettingsService} from "./settings.service"
import {IRootScopeService, ITimeoutService} from "angular"
import {getService, instantiateModule} from "../../../mocks/ng.mockhelper"
import {DEFAULT_SETTINGS, TEST_DELTA_MAP_A, TEST_DELTA_MAP_B} from "../util/dataMocks"
import {FileSelectionState, FileState, Settings} from "../codeCharta.model"
import {FileStateService} from "./fileState.service"
import {FileStateHelper} from "../util/fileStateHelper"
import {SettingsMerger} from "../util/settingsMerger"

describe("settingService", () => {
	let settingsService: SettingsService
	let $rootScope: IRootScopeService
	let $timeout: ITimeoutService

	let settings: Settings
	let fileStates: FileState[]

	beforeEach(() => {
		restartSystem()
		rebuildService()
		withMockedEventMethods()
	})

	function restartSystem() {
		instantiateModule("app.codeCharta.state")

		$rootScope = getService<IRootScopeService>("$rootScope")
		$timeout = getService<ITimeoutService>("$timeout")

		settings = JSON.parse(JSON.stringify(DEFAULT_SETTINGS))
		fileStates = [{ file: JSON.parse(JSON.stringify(TEST_DELTA_MAP_A)), selectedAs: FileSelectionState.Comparison },
			{ file: JSON.parse(JSON.stringify(TEST_DELTA_MAP_B)), selectedAs: FileSelectionState.Reference }]
	}

	function rebuildService() {
		settingsService = new SettingsService($rootScope, $timeout)
	}

	function withMockedEventMethods() {
		$rootScope.$on = settingsService["$rootScope"].$on = jest.fn()
	}

	describe("constructor", () => {
		it("should set settings to default settings", () => {
			rebuildService()

			expect(settingsService.getSettings()).toEqual(settings)
		})

		it("should subscribe to FileStateService", () => {
			FileStateService.subscribe = jest.fn()

			rebuildService()

			expect(FileStateService.subscribe).toHaveBeenCalledWith($rootScope, settingsService)
		})
	})

	describe("onFileSelectionStateChanged", () => {
		beforeEach(() => {
			settingsService.updateSettings = jest.fn()
			FileStateHelper.isPartialState = jest.fn().mockReturnValue(false)
			FileStateHelper.getVisibleFileStates = jest.fn().mockReturnValue(fileStates)
			SettingsMerger.getMergedFileSettings = jest.fn().mockReturnValue(DEFAULT_SETTINGS)
		})

		it("should call updateSettings with newFileSettings", () => {
			settingsService.onFileSelectionStatesChanged(fileStates, undefined)

			expect(settingsService.updateSettings).toHaveBeenCalledWith({fileSettings: DEFAULT_SETTINGS})
		})

		it("should call isPartialState with fileStates", () => {
			settingsService.onFileSelectionStatesChanged(fileStates, undefined)

			expect(FileStateHelper.isPartialState).toHaveBeenCalledWith(fileStates)
		})

		it("should call getVisibleFileStates with fileStates", () => {
			settingsService.onFileSelectionStatesChanged(fileStates, undefined)

			expect(FileStateHelper.getVisibleFileStates).toHaveBeenCalledWith(fileStates)
		})

		it("should call getMergedFileStates with visibleFiles and withUpdatedPath", () => {
			settingsService.onFileSelectionStatesChanged(fileStates, undefined)
			const visibleFiles = [fileStates[0].file, fileStates[1].file]

			expect(SettingsMerger.getMergedFileSettings).toHaveBeenCalledWith(visibleFiles, false)
		})
	})

	describe("updateSettings", () => {
		it("should set settings correctly", () => {
			const expected = {...DEFAULT_SETTINGS, appSettings: { ...DEFAULT_SETTINGS.appSettings, invertHeight: true } }

			settingsService.updateSettings({appSettings : { invertHeight: true }})

			expect(settingsService.getSettings()).toEqual(expected)
		})
	})

	describe("getDefaultSettings", () => {
		it("should match snapshot", () => {
			expect(settingsService.getDefaultSettings()).toMatchSnapshot()
		})
	})

	describe("subscribe", () => {
		it("should setup one event listener", () => {
			SettingsService.subscribe($rootScope, undefined)

			expect($rootScope.$on).toHaveBeenCalledTimes(1)
		})
	})
})