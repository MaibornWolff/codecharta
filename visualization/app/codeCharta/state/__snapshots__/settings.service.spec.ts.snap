// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`settingService getDefaultSettings should match snapshot 1`] = `
Object {
  "appSettings": Object {
    "amountOfTopLabels": 1,
    "camera": Vector3 {
      "x": 0,
      "y": 300,
      "z": 1000,
    },
    "deltaColorFlipped": false,
    "dynamicMargin": true,
    "enableEdgeArrows": true,
    "hideFlatBuildings": true,
    "invertHeight": false,
    "isWhiteBackground": false,
    "mapColors": Object {
      "angularGreen": "#00BFA5",
      "base": "#666666",
      "defaultC": "#89ACB4",
      "flat": "#AAAAAA",
      "lightGrey": "#DDDDDD",
      "markingColors": Array [
        "#FF1D8E",
        "#1d8eff",
        "#1DFFFF",
        "#8eff1d",
        "#8e1dff",
        "#FFFF1D",
      ],
      "negative": "#820E0E",
      "negativeDelta": "#ff0E0E",
      "neutral": "#ddcc00",
      "positive": "#69AE40",
      "positiveDelta": "#69FF40",
      "selected": "#EB8319",
    },
    "maximizeDetailPanel": false,
    "scaling": Vector3 {
      "x": 1,
      "y": 1,
      "z": 1,
    },
    "whiteColorBuildings": false,
  },
  "dynamicSettings": Object {
    "areaMetric": null,
    "colorMetric": null,
    "focusedNodePath": "",
    "heightMetric": null,
    "margin": null,
    "neutralColorRange": Object {
      "flipped": false,
      "from": null,
      "to": null,
    },
    "searchPattern": "",
    "searchedNodePaths": Array [],
  },
  "fileSettings": Object {
    "attributeTypes": Object {},
    "blacklist": Array [],
    "edges": Array [],
    "markedPackages": Array [],
  },
  "treeMapSettings": Object {
    "mapSize": 500,
  },
}
`;
